lib/gringotts/gateways/sagepay.ex

```diff
@@ -0,0 +1,362 @@
+    # commit(args, ...)
+  end
+
+  ###############################################################################
+  #                                PRIVATE METHODS                              #
+  ###############################################################################
+
+  # Makes the request to sagepay's network.
+  # For consistency with other gateway implementations, make your (final)
+  # network request in here, and parse it using another private method called
+  # `respond`.
+
+  # @spec commit(_) :: {:ok | :error, Response}
+  def commit(:post, endpoint, params, opts) do
+    a_url = @url <> endpoint
+    response = HTTPoison.post(a_url, params, opts)
+    format_response(response)
+  end
+
+  # Function generate_mkey generate a merchant_session_key that will exist only for 400 seconds
+  # and for 3 wrong card identifiers
+
+  defp generate_mkey(opts) do
+    vender_name = Poison.encode!(%{vendorName: opts[:config].vendor})
+
+    merchant_header = [
+      {"Authorization", "Basic " <> opts[:config].auth_id},
+      {"Content-type", "application/json"}
+    ]
+
+    key =
+      commit(:post, "merchant-session-keys", vender_name, merchant_header)
+      |> Map.get("merchantSessionKey")
+  end
+
+  # Function generate_ckey generate a unique cardidentifier for every transaction
+
+  defp generate_ckey(card, merchant_key) do
+    card_header = [
+      {"Authorization", "Bearer " <> merchant_key},
+      {"Content-type", "application/json"}
+    ]
+
+    card = card |> Poison.encode!()
+
+    card_identifier =
+      commit(:post, "card-identifiers", card, card_header)
+      |> Map.get("cardIdentifier")
+  end
+
+  # Function card_details creates the actual body (details of the customer )of the card 
+  # and with merchant_session_key, card_identifiier ,shipping address of a
+  # customer, and other details and converting the map into keyword list
+
+  defp card_details(amount, merchant_key, card_key, opts) do
+    card_body =
+      Poison.encode!(%{
+        "transactionType" => opts[:transactionType],
+        "paymentMethod" => %{
+          "card" => %{
+            "merchantSessionKey" => merchant_key,
+            "cardIdentifier" => card_key,
+            "save" => false
+          }
+        },
+        "vendorTxCode" => opts[:vendorTxCode],
+        "amount" => amount,
+        "currency" => opts[:currency],
+        "description" => opts[:description],
+        "customerFirstName" => opts[:customerFirstName],
+        "customerLastName" => opts[:customerLastName],
+        "billingAddress" => %{
+          "address1" => opts[:billingAddress].address1,
+          "city" => opts[:billingAddress].city,
+          "postalCode" => opts[:billingAddress].postalCode,
+          "country" => opts[:billingAddress].country
+        }
+      })
+  end
+
+  # Parses sagepay's response and returns a `Gringotts.Response` struct
+  # in a `:ok`, `:error` tuple.
+
+  defp format_response(response) do
```

Comment:
1. This function does not return a `Response` struct, though `@spec` of `authorize` suggests that it would return a `Response` tuple.
2. Please use one of the `respond` clauses from below.